pkg/targetgroupbinding/targets_manager.go

```diff
@@ -0,0 +1,276 @@
+package targetgroupbinding
+
+import (
+	"context"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	elbv2sdk "github.com/aws/aws-sdk-go/service/elbv2"
+	"k8s.io/apimachinery/pkg/util/cache"
+	"sigs.k8s.io/aws-alb-ingress-controller/pkg/aws/services"
+	"sync"
+	"time"
+)
+
+const (
+	defaultTargetsCacheTTL            = 10 * time.Minute
+	defaultRegisterTargetsChunkSize   = 200
+	defaultDeregisterTargetsChunkSize = 200
+)
+
+// TargetsManager is an abstraction around ELBV2's targets API.
+type TargetsManager interface {
+	// Register Targets into TargetGroup.
+	RegisterTargets(ctx context.Context, tgARN string, targets []elbv2sdk.TargetDescription) error
+
+	// Deregister Targets from TargetGroup.
+	DeregisterTargets(ctx context.Context, tgARN string, targets []elbv2sdk.TargetDescription) error
+
+	// List Targets from TargetGroup.
+	ListTargets(ctx context.Context, tgARN string) ([]TargetInfo, error)
+}
+
+// NewCachedTargetsManager constructs new cachedTargetsManager
+func NewCachedTargetsManager(elbv2Client services.ELBV2) *cachedTargetsManager {
+	return &cachedTargetsManager{
+		elbv2Client:                elbv2Client,
+		targetsCache:               cache.NewExpiring(),
+		targetsCacheTTL:            defaultTargetsCacheTTL,
+		registerTargetsChunkSize:   defaultRegisterTargetsChunkSize,
+		deregisterTargetsChunkSize: defaultDeregisterTargetsChunkSize,
+	}
+}
+
+var _ TargetsManager = &cachedTargetsManager{}
+
+// an cached implementation for TargetsManager.
+// Targets for each TargetGroup will be refreshed per targetsCacheTTL.
+// When list Targets with RefreshTargets list Option set,
+// only targets with ongoing TargetHealth(unknown/initial/draining) TargetHealth will be refreshed.
```

Comment:
IsOngoing() returns true only for initial/draining. Which one is correct? or ongoing here means something else?

Reply:
ongoing means targets in initial/draining. These targets's targetHealth might change, so need to be refreshed.

I changed the implementation to refresh "not healthy" targets instead of "ongoing" targets, to give "unhealthy" targets a chance to turn into "healthy".